File: src/lib/planning-stream.ts
import type { PlanningProgressEvent, ReasoningEffort } from "../types";
import {
  classifyUnsupportedReasoning,
  prepareReasoningRequest,
  readUpstreamHttpError,
  reasoningFallbackNotice,
  rememberUnsupportedReasoning,
} from "./reasoning-effort";

const MAX_PLANNING_RESPONSE_BYTES = 2 * 1024 * 1024;

interface PlanningRequestOptions {
  endpoint: string;
  apiKey: string;
  model: string;
  reasoningEffort: ReasoningEffort;
  body: Record<string, unknown>;
  requestId: string;
  signal: AbortSignal;
  onProgress?: (event: PlanningProgressEvent) => void;
  operation: string;
}

/**
 * 规划请求使用普通 JSON 响应。规划结果需要完整读取后再校验，避免把半截 JSON
 * 或服务商推理字段误展示为用户可见内容。
 */
export async function requestPlanningContent(
  options: PlanningRequestOptions,
): Promise<string> {
  emit(options, { type: "request-started", requestId: options.requestId });
  const attempt = prepareReasoningRequest(
    options.body,
    options.endpoint,
    options.model,
    options.reasoningEffort,
  );
  if (attempt.fallbackNotice) {
    emit(options, {
      type: "fallback-reasoning-auto",
      requestId: options.requestId,
      detail: attempt.fallbackNotice,
    });
  }

  try {
    return await requestPlanningAttempt({ ...options, body: attempt.body });
  } catch (error) {
    if (!attempt.appliedEffort) throw error;
    const scope = classifyUnsupportedReasoning(error, attempt.appliedEffort);
    if (!scope) throw error;
    rememberUnsupportedReasoning(
      options.endpoint,
      options.model,
      attempt.appliedEffort,
      scope,
    );
    emit(options, {
      type: "fallback-reasoning-auto",
      requestId: options.requestId,
      detail: reasoningFallbackNotice(attempt.appliedEffort),
    });
    return requestPlanningAttempt(options);
  }
}

async function requestPlanningAttempt(
  options: PlanningRequestOptions,
): Promise<string> {
  const response = await fetch(options.endpoint, {
    method: "POST",
    headers: requestHeaders(options.apiKey),
    body: JSON.stringify(options.body),
    signal: options.signal,
  });
  if (!response.ok) {
    throw await readUpstreamHttpError(response, options.operation, options.apiKey);
  }
  emit(options, { type: "response-started", requestId: options.requestId });
  return readJsonPlanningResponse(response, options);
}

async function readJsonPlanningResponse(
  response: Response,
  options: PlanningRequestOptions,
): Promise<string> {
  const text = await readPlanningResponseText(response);
  try {
    const payload = JSON.parse(text) as {
      choices?: Array<{ message?: unknown }>;
    };
    const message = asRecord(payload.choices?.[0]?.message);
    const content = readMessageContent(message?.content);
    if (!content.trim()) {
      throw new Error(`${options.operation}未返回规划内容。`);
    }
    emit(options, {
      type: "content-delta",
      requestId: options.requestId,
      receivedChars: Array.from(content).length,
    });
    return content;
  } catch (error) {
    if (error instanceof Error && error.message.includes("未返回规划内容")) {
      throw error;
    }
    throw new Error(`${options.operation}返回的 JSON 无法解析。`);
  }
}

async function readPlanningResponseText(response: Response): Promise<string> {
  const advertisedLength = Number(response.headers.get("content-length"));
  if (Number.isFinite(advertisedLength) && advertisedLength > MAX_PLANNING_RESPONSE_BYTES) {
    throw new Error("AI 规划返回的数据过大，已停止读取。");
  }
  if (!response.body) return "";

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let receivedBytes = 0;
  let text = "";
  while (true) {
    const chunk = await reader.read();
    receivedBytes += chunk.value?.byteLength ?? 0;
    if (receivedBytes > MAX_PLANNING_RESPONSE_BYTES) {
      await reader.cancel();
      throw new Error("AI 规划返回的数据过大，已停止读取。");
    }
    text += decoder.decode(chunk.value, { stream: !chunk.done });
    if (chunk.done) return text;
  }
}

function readMessageContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((part) => part && typeof part === "object" && "text" in part
        ? String((part as { text?: unknown }).text ?? "")
        : "")
      .join("\n");
  }
  return "";
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object"
    ? value as Record<string, unknown>
    : undefined;
}

function requestHeaders(apiKey: string): HeadersInit {
  return {
    Authorization: `Bearer ${apiKey}`,
    "Content-Type": "application/json",
    Accept: "application/json",
  };
}

function emit(
  options: PlanningRequestOptions,
  event: PlanningProgressEvent,
): void {
  options.onProgress?.(event);
}

File: src-tauri/src/planning_stream.rs
use reqwest::{Client, Response, StatusCode};
use serde::Serialize;
use serde_json::Value;
use tauri::ipc::Channel;

use crate::{
    http::read_limited,
    reasoning::{
        classify_unsupported_reasoning, prepare_reasoning_request, reasoning_fallback_notice,
        remember_unsupported_reasoning, upstream_error_detail,
    },
    types::CommandError,
};

const MAX_RESPONSE_BYTES: usize = 2 * 1024 * 1024;

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanningProgressEvent {
    #[serde(rename = "type")]
    event_type: String,
    request_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    received_chars: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    detail: Option<String>,
}

pub struct PlanningRequest<'a> {
    pub client: &'a Client,
    pub endpoint: &'a str,
    pub api_key: &'a str,
    pub model: &'a str,
    pub reasoning_effort: Option<&'a str>,
    pub request_id: &'a str,
    pub body: &'a Value,
    pub operation: &'a str,
    pub error_code: &'a str,
    pub on_event: &'a Channel<PlanningProgressEvent>,
}

pub async fn request_content(request: PlanningRequest<'_>) -> Result<String, CommandError> {
    emit(&request, "request-started", None, None);
    let reasoning_attempt = prepare_reasoning_request(
        request.body,
        request.endpoint,
        request.model,
        request.reasoning_effort,
    )?;
    if let Some(notice) = reasoning_attempt.fallback_notice.as_deref() {
        emit(&request, "fallback-reasoning-auto", None, Some(notice));
    }

    let result = request_content_attempt(&request, &reasoning_attempt.body).await;
    let Err(error) = result else {
        return result;
    };
    let Some(effort) = reasoning_attempt.applied_effort.as_deref() else {
        return Err(error);
    };
    let Some(scope) = classify_unsupported_reasoning(error.status, &error.message, effort) else {
        return Err(error);
    };
    remember_unsupported_reasoning(request.endpoint, request.model, effort, scope);
    let notice = reasoning_fallback_notice(effort);
    emit(&request, "fallback-reasoning-auto", None, Some(&notice));
    request_content_attempt(&request, request.body).await
}

async fn request_content_attempt(
    request: &PlanningRequest<'_>,
    body: &Value,
) -> Result<String, CommandError> {
    let response = request
        .client
        .post(request.endpoint)
        .bearer_auth(request.api_key)
        .json(body)
        .send()
        .await
        .map_err(|_| network_error(request.endpoint))?;

    if !response.status().is_success() {
        let status = response.status();
        let bytes = read_limited(response, MAX_RESPONSE_BYTES, "RESPONSE_TOO_LARGE").await?;
        return Err(upstream_error(
            request,
            status,
            upstream_error_detail(&bytes, request.api_key),
        ));
    }

    emit(request, "response-started", None, None);
    read_json_response(response, request).await
}

async fn read_json_response(
    response: Response,
    request: &PlanningRequest<'_>,
) -> Result<String, CommandError> {
    let bytes = read_limited(response, MAX_RESPONSE_BYTES, "RESPONSE_TOO_LARGE").await?;
    let payload = serde_json::from_slice::<Value>(&bytes).map_err(|_| {
        CommandError::new(
            format!("{}返回的 JSON 无法解析。", request.operation),
            "INVALID_PLANNING_RESPONSE",
        )
    })?;
    let content = message_content(&payload).ok_or_else(|| {
        CommandError::new(
            format!("{}未返回规划内容。", request.operation),
            "MISSING_PLANNING_RESULT",
        )
    })?;
    emit(
        request,
        "content-delta",
        Some(content.chars().count()),
        None,
    );
    Ok(content)
}

fn message_content(payload: &Value) -> Option<String> {
    response_message(payload)?
        .get("content")
        .and_then(content_value)
        .filter(|content| !content.trim().is_empty())
}

fn response_message(payload: &Value) -> Option<&Value> {
    payload.get("choices")?.as_array()?.first()?.get("message")
}

fn content_value(value: &Value) -> Option<String> {
    if let Some(text) = value.as_str() {
        return Some(text.to_owned());
    }
    let parts = value.as_array()?;
    Some(
        parts
            .iter()
            .filter_map(|part| part.get("text").and_then(Value::as_str))
            .collect::<Vec<_>>()
            .join("\n"),
    )
}

fn upstream_error(
    request: &PlanningRequest<'_>,
    status: StatusCode,
    detail: String,
) -> CommandError {
    let message = if detail.is_empty() {
        format!("{}失败（HTTP {}）。", request.operation, status.as_u16())
    } else {
        detail
    };
    CommandError::with_status(message, request.error_code, status.as_u16())
}

fn emit(
    request: &PlanningRequest<'_>,
    event_type: &str,
    received_chars: Option<usize>,
    detail: Option<&str>,
) {
    let _ = request.on_event.send(PlanningProgressEvent {
        event_type: event_type.to_owned(),
        request_id: request.request_id.to_owned(),
        received_chars,
        detail: detail.map(str::to_owned),
    });
}

fn network_error(endpoint: &str) -> CommandError {
    CommandError::new(
        format!("无法连接目标 API（{}）。", endpoint),
        "NETWORK_ERROR",
    )
}
